### PR TITLE
Wind Downscaling CLI

### PR DESCRIPTION
**Pull request by meabhnicguidhir**
Issue #67 
- CLI written
- sample input and known good output files created for bats tests.
- Bats tests for: no arguments, -help option, acceptance test with all compulsory arguments, acceptance test with -vegetation cover optional field.

Notes:
- currently target orography = standard orography. When target orography is defined in the future, acceptance tests will have to be updated.
- some technical debt notes were made on the original issue ( #67 ). These should be addressed in technical debt issues.

Testing:
 - [x] Ran tests and they passed OK
 - [x] Added new tests for the new feature(s)

### CODE REVIEW

**Comment by benfitzpatrick** [comment]
This looks good! When you say target orography == standard orography for acceptance tests, does that mean we're using the orography derived from the overly smoothed orography from the model? Would you like me to dig out a PP orography from somewhere? I guess there isn't much height correction to do currently.

File: tests/improver-wind-downscaling/02_basic.bats [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+. $IMPROVER_DIR/tests/lib/utils
+
+@test "wind downscaling wind_speed " {
+  TEST_DIR=$(mktemp -d)
+  improver_check_skip_acceptance
+  test_path="$IMPROVER_ACC_TEST_DIR/wind_downscaling/basic/"
+
+  # Run wind downscaling processing and check it passes.
+  run improver wind-downscaling "$test_path/input.nc" "$test_path/a_over_s.nc" \
```
Comment (benfitzpatrick) [inline tests/improver-wind-downscaling/02_basic.bats]: Are you using a similar input file to the percentile wind speed, with realization as the first dimension?

Comment (meabhnicguidhir) [inline tests/improver-wind-downscaling/02_basic.bats]: No, the input isn't very similar to the percentile input. My input file is a wind field on height levels with no realization component and time as a scalar coordinate.  The percentile field is a 10m wind field with realization as the 3rd dimension coordinate.

**Comment by meabhnicguidhir** [comment]
Hi Ben,
Yes, currently the acceptance tests use the regridded model orography as both the source orography and the target orography we are downscaling to, which is obviously not ideal. If you could get a pp orography field onto the ukvx grid, that would be great!

File: bin/improver-wind-downscaling [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+    parser.add_argument('model_resolution', metavar='MODEL_RESOLUTION',
+                        help='Original resolution of model orography (before'
+                             ' interpolation to standard grid).'
+                             ' Units of field: m')
+    parser.add_argument('output_filepath', metavar='OUTPUT_FILE',
+                        help='The output path for the processed NetCDF')
+    parser.add_argument('--height_levels_filepath', metavar='HEIGHT_LEVELS',
+                        help='Location of file containing height levels '
+                             'coincident with wind speed field.')
+    parser.add_argument('--z0_filepath', metavar='VEGETATIVE_ROUGHNESS_LENGTH',
```
Comment (fionaRust) [inline bin/improver-wind-downscaling]: Is `z0_filepath` the best name for this?

File: bin/improver-wind-downscaling [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+def main():
+    """Load in arguments and get going."""
+    parser = argparse.ArgumentParser(
+        description='Run wind downscaling to apply roughness correction and'
+                    ' height correction to wind fields (as described in'
+                    ' Howard and Clark [2007]).')
+    parser.add_argument('wind_speed_filepath', metavar='WIND_SPEED_FILE',
+                        help='Location of the wind speed on standard grid'
+                             ' file. Any units can be supplied.')
+    parser.add_argument('a_over_s_filepath',  metavar='AOS_FILE',
```
Comment (fionaRust) [inline bin/improver-wind-downscaling]: Is `a_over_s_filepath` the best name for this?

File: lib/improver/wind_downscaling.py [hunk 4]
```diff
@@ -734,7 +734,7 @@ def calc_av_ppgrid_res(self, a_cube):
         [exp_xname, exp_yname] = ["projection_x_coordinate",
                                   "projection_y_coordinate"]
         exp_unit = Unit("m")
-        if (x_name is not exp_xname) or (y_name is not exp_yname):
+        if (x_name != exp_xname) or (y_name != exp_yname):
```
Comment (fionaRust) [inline lib/improver/wind_downscaling.py]: Why is this change needed?

File: tests/improver-wind-downscaling/01_help.bats [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+positional arguments:
+  WIND_SPEED_FILE       Location of the wind speed on standard grid file. Any
+                        units can be supplied.
+  AOS_FILE              Location of model silhouette roughness file. Units of
+                        field: dimensionless
+  SIGMA_FILE            Location of standard deviation of model orography
+                        height file. Units of field: m
+  TARGET_OROGRAPHY_FILE
+                        Location of target orography file to downscale fields
+                        to.Units of field: m
```
Comment (fionaRust) [inline tests/improver-wind-downscaling/01_help.bats]: Missing a space here, which should probably be corrected to make things clearer

File: tests/improver-wind-downscaling/01_help.bats [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+  STANDARD_OROGRAPHY_FILE
+                        Location of orography on standard grid file
+                        (interpolated model orography. Units of field: m
```
Comment (fionaRust) [inline tests/improver-wind-downscaling/01_help.bats]: Missing a closed bracket

**Comment by fionaRust** [comment]
Have you looked at the known good output you are adding? Have you plotted it to see if it looks ok?

File: lib/improver/wind_downscaling.py (hunk 4 above)
Comment (meabhnicguidhir) [inline lib/improver/wind_downscaling.py]: The previous version was entering the if loop even when the expected names are found. It seems that 'is not' checks the identities are not the same, whereas != tests for inequality between the strings. I think that because x_name comes from the name() method being called on the coordinate, it has a different 'identity' to the string exp_xname, even though the strings are equal. The problem in the original code is easy to create on the command line, for ex:
```
> test = cube.coord(axis='x')[0].name()
> print test
projection_x_coordinate
> test=='projection_x_coordinate'
True
> test is 'projection_x_coordinate'
False
```

**Comment by meabhnicguidhir** [comment]
I have looked briefly at the output I've added. It looks fairly similar to the input, with some small differences, particularly in mountaineous areas which I think is about right.

File: bin/improver-wind-downscaling (hunk 2 above)
Comment (meabhnicguidhir) [inline bin/improver-wind-downscaling]: changed to veg_roughness_filepath

File: bin/improver-wind-downscaling (hunk 3 above)
Comment (meabhnicguidhir) [inline bin/improver-wind-downscaling]: changed to silhouette_roughness_filepath

File: tests/improver-wind-downscaling/01_help.bats (hunk 5 above)
Comment (meabhnicguidhir) [inline tests/improver-wind-downscaling/01_help.bats]: Added!

File: tests/improver-wind-downscaling/01_help.bats (hunk 6 above)
Comment (meabhnicguidhir) [inline tests/improver-wind-downscaling/01_help.bats]: Added!

File: lib/improver/wind_downscaling.py (hunk 4 above)
Comment (bayliffe) [inline lib/improver/wind_downscaling.py]: Ooo, interesting. :whale2:

File: bin/improver-wind-downscaling [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+                        help='Original resolution of model orography (before'
+                             ' interpolation to standard grid).'
+                             ' Units of field: m')
+    parser.add_argument('output_filepath', metavar='OUTPUT_FILE',
+                        help='The output path for the processed NetCDF')
+    parser.add_argument('--height_levels_filepath', metavar='HEIGHT_LEVELS',
+                        help='Location of file containing height levels '
+                             'coincident with wind speed field.')
+    parser.add_argument('--veg_roughness_filepath',
+                        metavar='VEGETATIVE_ROUGHNESS_LENGTH',
```
Comment (bayliffe) [inline bin/improver-wind-downscaling]: This is not in fact the VEGETATIVE_ROUGHNESS_LENGTH but rather a path to a file containing that information. The metavar should reflect this.

File: bin/improver-wind-downscaling [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+                        help='Location of orography on standard grid file '
+                             '(interpolated model orography.'
+                             ' Units of field: m')
+    parser.add_argument('model_resolution', metavar='MODEL_RESOLUTION',
+                        help='Original resolution of model orography (before'
+                             ' interpolation to standard grid).'
+                             ' Units of field: m')
+    parser.add_argument('output_filepath', metavar='OUTPUT_FILE',
+                        help='The output path for the processed NetCDF')
+    parser.add_argument('--height_levels_filepath', metavar='HEIGHT_LEVELS',
```
Comment (bayliffe) [inline bin/improver-wind-downscaling]: Not HEIGHT_LEVELS but instead a filepath to a file containing them, Should be reflected in metavar.

File: bin/improver-wind-downscaling [hunk 9]
```diff
@@ -0,0 +1,103 @@
...
+        args.silhouette_roughness_filepath)
+    sigma = iris.load_cube(args.sigma_filepath)
+    target_orog = iris.load_cube(args.target_orog_filepath)
+    standard_orog = iris.load_cube(args.standard_orog_filepath)
+    if args.height_levels_filepath:
+        height_levels = iris.load_cube(args.height_levels_filepath)
+    else:
+        height_levels = None
+    if args.veg_roughness_filepath:
+        veg_roughness_cube = iris.load_cube(args.veg_roughness_filepath)
```
Comment (bayliffe) [inline bin/improver-wind-downscaling]: veg_roughness is my favourite name in Improver so far. :corn:

File: tests/improver-wind-downscaling/00_null.bats [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+@test "wind downscaling no arguments" {
+  run improver wind-downscaling
+  [[ "$status" -eq 2 ]]
+  read -d '' expected <<'__TEXT__' || true
+usage: improver-wind-downscaling [-h] [--height_levels_filepath HEIGHT_LEVELS]
+                                 [--veg_roughness_filepath VEGETATIVE_ROUGHNESS_LENGTH]
```
Comment (bayliffe) [inline tests/improver-wind-downscaling/00_null.bats]: Will need updating if names change in CLI.

File: tests/improver-wind-downscaling/01_help.bats [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+
+@test "wind downscaling -h" {
+  run improver wind-downscaling -h
+  [[ "$status" -eq 0 ]]
+  read -d '' expected <<'__HELP__' || true
+usage: improver-wind-downscaling [-h] [--height_levels_filepath HEIGHT_LEVELS]
+                                 [--veg_roughness_filepath VEGETATIVE_ROUGHNESS_LENGTH]
```
Comment (bayliffe) [inline tests/improver-wind-downscaling/01_help.bats]: Will also need changing if any names change in CLI.

File: tests/improver-wind-downscaling/03_veg.bats [hunk 12]
```diff
@@ -0,0 +1,51 @@
+#!/usr/bin/env bats
```
Comment (bayliffe) [inline tests/improver-wind-downscaling/03_veg.bats]: New favourite name :corn: :bat: :bat:

File: lib/improver/wind_downscaling.py (hunk 4 above)
Comment (fionaRust) [inline lib/improver/wind_downscaling.py]: Oooh ok. I'm happy with the change then.

File: bin/improver-wind-downscaling (hunk 8 above)
Comment (meabhnicguidhir) [inline bin/improver-wind-downscaling]: True, it should. Changed.

File: bin/improver-wind-downscaling (hunk 7 above)
Comment (meabhnicguidhir) [inline bin/improver-wind-downscaling]: Also true, it also should. Also changed.

File: bin/improver-wind-downscaling (hunk 9 above)
Comment (meabhnicguidhir) [inline bin/improver-wind-downscaling]: :smile_cat:

File: tests/improver-wind-downscaling/00_null.bats (hunk 10 above)
Comment (meabhnicguidhir) [inline tests/improver-wind-downscaling/00_null.bats]: Updated this test.

File: tests/improver-wind-downscaling/01_help.bats (hunk 11 above)
Comment (meabhnicguidhir) [inline tests/improver-wind-downscaling/01_help.bats]: Also updated this test.

File: tests/improver-wind-downscaling/03_veg.bats (hunk 12 above)
Comment (meabhnicguidhir) [inline tests/improver-wind-downscaling/03_veg.bats]: If only there was a fruit roughness correction.

File: tests/improver-wind-downscaling/02_basic.bats (hunk 1 above)
Comment (benfitzpatrick) [inline tests/improver-wind-downscaling/02_basic.bats]: Sorry for the delay - could you please alter this so that it is capable of taking in a realization-first-dimension cube (or can only take one in)? This is how we intend to use it in the simple suite.

Comment (meabhnicguidhir) [inline tests/improver-wind-downscaling/02_basic.bats]: I agree this is needed, but think it is out of scope for this ticket, as the current implementation assumes a 4D or 3D cube with x,y,z (and optional t) dimensions. This problem has been added to the list of potential technical debt issues, see: https://github.com/metoppv/improver/issues/67#issuecomment-308048647)

File: bin/improver-wind-downscaling [hunk 13]
```diff
@@ -0,0 +1,104 @@
...
+                        metavar='TARGET_OROGRAPHY_FILE',
+                        help='Location of target orography file to downscale'
+                             ' fields to.'
+                             'Units of field: m')
+    parser.add_argument('standard_orog_filepath',
+                        metavar='STANDARD_OROGRAPHY_FILE',
+                        help='Location of orography on standard grid file '
+                             '(interpolated model orography.'
+                             ' Units of field: m')
+    parser.add_argument('model_resolution', metavar='MODEL_RESOLUTION',
```
Comment (fionaRust) [inline bin/improver-wind-downscaling]: Just a question, how does this deal with the variable resolution over the model domain?

File: bin/improver-wind-downscaling [hunk 14]
```diff
@@ -0,0 +1,104 @@
...
+
+import iris
+
+from improver import wind_downscaling
+
+
+def main():
+    """Load in arguments and get going."""
+    parser = argparse.ArgumentParser(
+        description='Run wind downscaling to apply roughness correction and'
```
Comment (fionaRust) [inline bin/improver-wind-downscaling]: Is it true all the inputs need to be on the same grid? If so maybe this should be stated, as it wouldn't be immediately obvious that the sigma file would be on the same grid as everything else.

**Comment by gavinevans** [comment]
Follow-on tickets, as discussed in https://github.com/metoppv/improver/issues/67#issuecomment-308048647, have been raised: #121, #122.

File: bin/improver-wind-downscaling (hunk 13 above)
Comment (meabhnicguidhir) [inline bin/improver-wind-downscaling]: Currently, you can only pass in one value, so it can't deal with variable resolution. It uses this value to calculate the wavenumber of  typical orographic lengthscale, which is used in the height correction. We've tested using 2200m, which is the value @gavinevans used for the Mogreps-UK demo. For now will we be able to make do using the resolution of the central, fixed resolution part of the domain, as we have tested with?

File: bin/improver-wind-downscaling (hunk 14 above)
Comment (meabhnicguidhir) [inline bin/improver-wind-downscaling]: Yes, they all have to be on the same grid. I've added this to the description in the cli.

File: bin/improver-wind-downscaling (hunk 7 above)
Comment (bayliffe) [inline bin/improver-wind-downscaling]: Also, thank you.

File: lib/improver/wind_downscaling.py (hunk 4 above)
Comment (bayliffe) [inline lib/improver/wind_downscaling.py]: :whale:

File: bin/improver-wind-downscaling [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        veg_roughness_cube = iris.load_cube(args.veg_roughness_filepath)
+    else:
+        veg_roughness_cube = None
+    try:
+        realization_coord = wind_speed.coord('realization')
+        wind_speed_iterator = wind_speed.slices_over('realization')
+    except CoordinateNotFoundError:
+        wind_speed_iterator = [wind_speed]
+    wind_speed_list = iris.cube.CubeList()
+    for wind_speed_slice in wind_speed_iterator:
```
Comment (benfitzpatrick) [inline bin/improver-wind-downscaling]: Nice, I like this - as discussed, we can multiprocess this out in future if we need to.

**Comment by fionaRust** [comment]
The known good output has different dimensions to the input cube, the realisation seems to be a scalar coord (as there is only one). Maybe we should look at this.

**Comment by benfitzpatrick** [comment]
That is a bit of a problem, since we know that we might use single valued realization dimensions for the deterministic processing.

**Comment by benfitzpatrick** [comment]
I'd suggest using http://scitools.org.uk/iris/docs/latest/iris/iris/util.html#iris.util.new_axis to repromote it

File: tests/improver-wind-downscaling/02_basic.bats (hunk 1 above)
Comment (fionaRust) [inline tests/improver-wind-downscaling/02_basic.bats]: This has been implemented by looping through realizations in the CLI.
